pysat/utils/registry.py

```diff
@@ -15,117 +15,296 @@
+
+        # second option, use a small mod to the test method signature
+        validate.test_modules_standard('', inst_module)
+        validate.test_standard_function_presence('', inst_module)
+
+        # registry is a dict of dicts
+        # platform, name, module string
+        # get platform and name identifiers from imported module
+        platform = inst_module.platform
+        name = inst_module.name
+        # only register module if not already present
+        # multiple names allowed for a single platform
+        if platform not in pysat.user_modules:
+            # setup `of dict` part of dict of dicts
+            pysat.user_modules[platform] = {}
+        # only register name if not present under platform
+        if name not in pysat.user_modules[platform]:
+            logger.info('Registering user module {}'.format(module_name))
+            # add to current user modules structure
+            pysat.user_modules[platform][name] = module_name
+            # store user modules to disk
+            store()
+        else:
+            # platform/name combination already registered
+            estr = ' '.join(('An instrument has already been registered using',
+                             'platform:', platform, 'and name:', name,
+                             'which maps to:', module_name))
+            raise ValueError(estr)
 
+    return
 
-    testInst = Instrument()
 
+def register_by_module(module):
+    """Register all sub-modules attached to input module
+
+    Enables instantiation of a third-party Instrument
+    module using
+    ::
+        inst = pysat.Instrument(platform, name)
+
+    Parameters
+    ----------
+    module : Python module
+        Module with one or more pysat.Instrument support modules
+        attached as sub-modules to the input `module`
+
+    Raises
+    ------
+    ValueError
+        If platform and name associated with a module are already registered
 
+    Note
+    ----
+    Gets a list of sub-modules by using the __all__ attribute,
+    defined in the module's __init__.py
+
+    Examples
+    --------
+    ::
+        import pysat
+        import pysatModels
+        pysat.utils.registry.register_by_module(pysatModels.models)
+
+        import pysatSpaceWeather
+        pysat.utils.registry.register_by_module(pysatSpaceWeather.instruments)
 
     """
-    import importlib
-    inst_module = importlib.import_module(module_name)
 
-    if module_name not in pysat.user_modules:
-        print('registering user module {}'.format(module_name))
-        pysat.user_modules.append(module_name)
-        store()
+    # get a list of all user specified modules attached to module
+    module_names = module.__all__
+    # add in package preamble
+    module_names = [module.__name__ + '.' + mod for mod in module_names]
+    # register all of them
+    register(module_names)
 
-    return inst_module
+    return
 
 
-def remove(*module_names):
+def remove(platforms, names=None):
```

Comment:
Wouldn't it be cleaner to have `names` default to an empty list?  This would make the expected input clearer and not affect the functionality.

Reply:
I provided a suggestion in the docstring, but didn't make changes elsewhere.  It's really just the `is None` evaluation line that would need to be changed to `len(names) == 0` and also add a ValueError check for `len(names) != len(platforms)`.  This should probably be done anyhow.

Reply:
General guidance I've read is that objects that are modifiable in place shouldn't be used as defaults in the method signature. If names=[], or names='hi there!', is in the call signature and names gets modified in-place within the method, those changes will persist between calls since the signature definition code is only run at definition.

while it seems contrary to the whole point of the feature, I've seen the best-practices method signature is
```
def new_method(var1, keyword=None):
    if keyword is None:
        keyword = []
```

Reply:
>  objects that are modifiable in place shouldn't be used as defaults in the method signature

Yes, I am ok with `names` being an arg instead of a kwarg.  Is there any instance where someone wouldn't be expected to enter both?  It's required in every other area.

General guidance also recommends avoiding the use of `isinstance` for testing your input.  You can get around this by using `np.asarray` on the input.  This will be more robust in instances where someone has an input that for some reason isn't interpreted as a string (as has happened before and will likely happen again).

Reply:
>> Is there any instance where someone wouldn't be expected to enter both? It's required in every other area.

If a user wants to remove all instruments for a given platform(s). I like having a minimal number of required inputs but I am ok with changing names to an arg. I'll have something tomorrow.

 I do like to be able to support a single string or list-like of strings with the same code. I've tried using np.asarray and it hasn't worked out for me so far.
```
In [19]: for item in np.asarray('hello'): 
    ...:     print(item) 
    ...:                                                                                                                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-bff22eb1af27> in <module>
----> 1 for item in np.asarray('hello'):
      2     print(item)
      3 

TypeError: iteration over a 0-d array

```

I've tried duck typing multiple times on the string vs list issue. I've also tried checking for iter or other things like that. None are quite right so far.

Reply:
Switched to list-like for args platforms and names. Similar change to other methods that allowed a string or list like of strings, now just list-like.

Reply:
Cool.  For future reference, this was an alternative:
```
platforms = np.asarray(platforms)

if platforms.shape == ():
    platforms = np.array([platforms])
```

Reply:
Cool. I figured the registry wasn't a high user interaction area so probably not worth the potential future maintenance costs of fancy inputs. I'll remember this for areas that will see more use though!